core/src/main/java/org/mskcc/cbio/portal/dao/DaoGene.java

```diff
@@ -67,45 +67,58 @@ private static synchronized int getNextFakeEntrezId() throws DaoException {
     public static synchronized int addGeneWithoutEntrezGeneId(CanonicalGene gene) throws DaoException {
         CanonicalGene existingGene = getGene(gene.getHugoGeneSymbolAllCaps());
         gene.setEntrezGeneId(existingGene==null?getNextFakeEntrezId():existingGene.getEntrezGeneId());
-        return addGene(gene);
+        return addOrUpdateGene(gene);
     }
 
     /**
-     * Adds a new Gene Record to the Database.
+     * Adds a new Gene Record to the Database OR updates the given gene object
+     * with the geneticEntityId found in the DB for this gene.
+     * If it is a new gene, it will generate a new genetic entity id and
+     * update the given CanonicalGene with the generated 
+     * geneticEntityId. 
      *
      * @param gene Canonical Gene Object.
      * @return number of records successfully added.
      * @throws DaoException Database Error.
      */
-    public static int addGene(CanonicalGene gene) throws DaoException {
-        if (MySQLbulkLoader.isBulkLoad()) {
-            //  write to the temp file maintained by the MySQLbulkLoader
-            MySQLbulkLoader.getMySQLbulkLoader("gene").insertRecord(Long.toString(gene.getEntrezGeneId()),
-                    gene.getHugoGeneSymbolAllCaps(),gene.getType(),gene.getCytoband(),gene.getLength()==0?null:Integer.toString(gene.getLength()));
-            addGeneAliases(gene);
-            // return 1 because normal insert will return 1 if no error occurs
-            return 1;
-        }
-        
+    public static int addOrUpdateGene(CanonicalGene gene) throws DaoException {
+
         Connection con = null;
         PreparedStatement pstmt = null;
         ResultSet rs = null;
         try {
             int rows = 0;
             CanonicalGene existingGene = getGene(gene.getEntrezGeneId());
             if (existingGene == null) {
-                con = JdbcUtil.getDbConnection(DaoGene.class);
-                pstmt = con.prepareStatement
-                        ("INSERT INTO gene (`ENTREZ_GENE_ID`,`HUGO_GENE_SYMBOL`,`TYPE`,`CYTOBAND`,`LENGTH`) "
-                                + "VALUES (?,?,?,?,?)");
-                pstmt.setLong(1, gene.getEntrezGeneId());
-                pstmt.setString(2, gene.getHugoGeneSymbolAllCaps());
-                pstmt.setString(3, gene.getType());
-                pstmt.setString(4, gene.getCytoband());
-                pstmt.setInt(5, gene.getLength());
+            	//new gene, so add genetic entity first:
+            	int geneticEntityId = DaoGeneticEntity.addNewGeneticEntity(DaoGeneticEntity.EntityTypes.GENE);
```

Comment:
note: mysql autoincrement is what prohibits reuse of a newly generated geneticEntityId number.